src/prefect/utilities/local_cache.py

```diff
@@ -0,0 +1,195 @@
+import atexit
+import functools
+import inspect
+import threading
+import weakref
+import uuid
+from collections import defaultdict
+from typing import Any, Callable
+
+from toolz import curry
+
+import prefect
+from prefect.utilities.logging import get_logger
+
+
+__all__ = ("shared_object",)
+
+
+class _SerializableLock(object):
+    _locks = weakref.WeakValueDictionary()  # type: weakref.WeakValueDictionary
+    """A Serializable per-process Lock.
+
+    This wraps a normal ``threading.Lock`` or ``threading.RLock`` object and
+    satisfies the same interface.  However, this lock can also be serialized
+    and sent to different processes. It will not block concurrent operations
+    between processes but will consistently deserialize into the same lock.
+
+    This is useful for consistently protecting resources on a per-process
+    level.
+    """
+
+    def __init__(self, reentrant: bool = False, token: str = None):
+        self.reentrant = reentrant
+        self.token = token or str(uuid.uuid4())
+        if self.token in _SerializableLock._locks:
+            self.lock = _SerializableLock._locks[self.token]
+        else:
+            self.lock = threading.RLock() if reentrant else threading.Lock()
+            _SerializableLock._locks[self.token] = self.lock
+
+    def acquire(self, *args: Any, **kwargs: Any) -> bool:
+        return self.lock.acquire(*args, **kwargs)
+
+    def release(self, *args: Any, **kwargs: Any) -> None:
+        self.lock.release(*args, **kwargs)
+
+    def __enter__(self) -> None:
+        self.lock.__enter__()
+
+    def __exit__(self, *args: Any) -> None:
+        self.lock.__exit__(*args)
+
+    def __reduce__(self) -> tuple:
+        return _SerializableLock, (self.reentrant, self.token)
+
+    def __repr__(self) -> str:
+        return "<%s: %s>" % (self.__class__.__name__, self.token)
+
+
+@atexit.register
+def clear_all_local_caches() -> None:
+    """On process exit, ensure all local caches are cleared.
+
+    For most executors, this should be a no-op (the cleanup should have already run).
+    """
+    for flow_run_id in list(_CACHES):
+        clear_local_cache(flow_run_id)
+
+
+def _init_cache() -> dict:
+    out = defaultdict(dict)  # type: dict
+    out["generators"] = []
+    return out
+
+
+_CACHES = defaultdict(_init_cache)  # type: dict
```

Comment:
Whilst I can eventually figure out what is going to be stored inside `_CACHES` it takes looking at the code in a few different places. A brief comment here explaining the structure of this dict would be really helpful!